packages/dds/tree/docs/wip/identifiers.md

```diff
@@ -0,0 +1,73 @@
+# SharedTree Node Identifiers
+
+## Why Identifiers
+
+Why use identifiers at all? SharedTree has paths, are not those already sufficient "references" or "handles" to nodes in the tree? No, a node's identifier is related to, but not the same thing as, the path to that node in the tree. A path points to a location within a tree, but it does not necessarily capture any semantic meaning about the identity of the node it points to. Consider a node A which is deleted, and then consider a node B which is inserted in the same place. Node A and Node B have the same path (although they are in different revisions) but they are not the same node; they should therefore not share the same identifier. Likewise, consider a node A which is moved from location X to location Y in the tree. Node A has a different path after the move than it did before, but semantically it is the "same node"; it should have the same identifier in both locations.
+
+## Proposal Overview
+
+Identifiers are stored under nodes in fields. A SharedTree has an optional index, the IdentifierIndex. The IdentifierIndex allows a user to provide an identifier and retrieve a cursor to a node with that identifier. An identifier is a special type (a numerically compressed node identifier). The index maintains a mapping from each identifier to a path. If history is enabled, then each identifier maps to a list of tuples of [path, revision]. The index is parameterized over a field key, e.g. "identifier", and will consider all nodes that have this field to be identified by the data in that field. That data ought to be an identifier; if it is not, the index can either ignore it or fail.
```

Comment:
About "The IdentifierIndex allows a user to provide an identifier and retrieve a cursor to a node with that identifier.", would that not require the `IdentifierIndex` to have access to the forest? That seems like a dependency we may want to avoid since we want to make it possible to use ST (and possibly the `IdentifierIndex`) without a forest. Having the `IdentifierIndex` return a path seems fine though.

Reply:
That's a really good question. I'd been sort of assuming that we'd want the IdentifierIndex to have a ForestIndex passed to in on construction or something. I'm not sure what the use of looking up paths from identifiers is if you can't get the nodes that they refer to. That said, there might very well be a use case, and the architecture stays cleaner if we do as you suggested. We can always write helpers which take an ID and the two indexes and return the node rather than just the path.